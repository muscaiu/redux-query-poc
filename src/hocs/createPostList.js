import React, { PureComponent } from "react";
import { compose } from "redux";
import { connect } from "react-redux";
import {
  connectRequest,
  querySelectors
} from "redux-query";
import { getOr } from 'lodash/fp';

import { selectPosts } from "../selectors/postSelectors";
import { fetchPosts } from "../actions/postsActions";
import selectQueryStatus from '../lib/redux-query-utils';
import Spinner from '../components/Spinner';

const getQueryIsPending = getOr(true, 'isPending');

export default function (WrappedComponent) {
  class PostListHoc extends PureComponent {
    render() {
      if (this.props.isLoading) {
        return (<Spinner />)
      }
      return <WrappedComponent {...this.props} />;
    }
  }

  function mapStateToProps(state) {
    console.log('state:', state)
    const query = fetchPosts();
    const queryStatus = selectQueryStatus(state.queries, query);
    return {
      // queryStatus,
      // isLoading: selectQueryStatus(state.queries, query).isPending,
      // isLoading: querySelectors.isPending(state.queries, query),
      isLoading: getQueryIsPending(queryStatus),
      posts: selectPosts(state),
      query
    };
  }

  return compose(
    connect(mapStateToProps),
    connectRequest(props => props.query)
  )(PostListHoc);
}
