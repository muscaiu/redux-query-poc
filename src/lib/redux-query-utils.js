import { defaults } from 'lodash/fp';
import { getQueryKey } from 'redux-query';

const withDefaultStatus = defaults({
    isFinished: false,
    isPending: true,
    code: 0
});

export default function selectQueryStatus(state, queryConfig) {
    const queryKey = getQueryKey(queryConfig);
    const {
        isFinished,
        isPending,
        status: code,
        lastUpdated,
        queryCount
    } = state[queryKey] || {};

    return withDefaultStatus({
        isFinished,
        isPending,
        ...code && {
            code,
            lastUpdated,
            queryCount
        }
    });
}
