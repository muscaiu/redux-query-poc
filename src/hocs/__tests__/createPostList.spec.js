import React from 'react';
import { shallow } from 'enzyme';
import createStore from 'redux-mock-store';

import createPostList from '../createPostList';

const mockStore = createStore();
const state = {
    queries: {},
    entities: {}
};
describe('createPostList', () => {
    beforeEach(() => {
        jest.clearAllMocks();
    });

    it('should return a connected component', () => {
        const PostList = createPostList();
        const store = mockStore(state);
        const wrapper = shallow(<PostList />, { context: { store } });
        expect(wrapper).toMatchSnapshot();
    });

    it('it should return a Spinner', () => {
        const selectQueryStatus = jest.fn(() => ({ isPending: true }));
        require('../../lib/redux-query-utils').default = selectQueryStatus; // eslint-disable-line global-require

        const PostList = createPostList();
        const store = mockStore(state);
        const wrapper = shallow(<PostList />, { context: { store } });

        expect(selectQueryStatus).toHaveBeenCalled();
        expect(wrapper.props().isLoading).toBeTruthy();
        expect(wrapper.dive().dive().find('Spinner')).toHaveLength(1);
        expect(wrapper.dive().dive().find('Spinner')).toMatchSnapshot();
        // console.log(wrapper.debug());
        // console.log(wrapper.dive().debug());
        // console.log(wrapper.dive().dive().debug());
    });
})
