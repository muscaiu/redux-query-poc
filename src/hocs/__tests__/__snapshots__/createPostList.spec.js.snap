// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createPostList it should return a Spinner 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): ShallowWrapper {
    Symbol(enzyme.__root__): [Circular],
    Symbol(enzyme.__unrendered__): <PostListHoc
      dispatch={[Function]}
      forceRequest={[Function]}
      isLoading={true}
      posts={null}
      query={
            Object {
                  "body": undefined,
                  "force": undefined,
                  "meta": undefined,
                  "options": undefined,
                  "queryKey": undefined,
                  "retry": undefined,
                  "transform": [Function],
                  "type": "@@query/REQUEST_ASYNC",
                  "unstable_preDispatchCallback": undefined,
                  "update": Object {
                    "posts": [Function],
                  },
                  "url": "https://jsonplaceholder.typicode.com/posts",
                }
      }
/>,
    Symbol(enzyme.__renderer__): Object {
      "batchedUpdates": [Function],
      "checkPropTypes": [Function],
      "getNode": [Function],
      "render": [Function],
      "simulateError": [Function],
      "simulateEvent": [Function],
      "unmount": [Function],
    },
    Symbol(enzyme.__node__): Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {},
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    Symbol(enzyme.__nodes__): Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    Symbol(enzyme.__options__): Object {
      "adapter": ReactSixteenAdapter {
        "options": Object {
          "enableComponentDidUpdateOnSetState": true,
          "legacyContextMode": "parent",
          "lifecycles": Object {
            "componentDidUpdate": Object {
              "onSetState": true,
            },
            "getChildContext": Object {
              "calledByRenderer": false,
            },
            "getDerivedStateFromProps": true,
            "getSnapshotBeforeUpdate": true,
            "setState": Object {
              "skipsComponentDidUpdateOnNullish": true,
            },
          },
        },
      },
      "context": Object {
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
      },
    },
  },
  Symbol(enzyme.__unrendered__): null,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {},
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {},
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "context": Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
    },
  },
  Symbol(enzyme.__rootNodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {},
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
}
`;

exports[`createPostList should return a connected component 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Connect(ConnectRequest(PostListHoc)) />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "dispatch": [Function],
      "isLoading": true,
      "posts": null,
      "query": Object {
        "body": undefined,
        "force": undefined,
        "meta": undefined,
        "options": undefined,
        "queryKey": undefined,
        "retry": undefined,
        "transform": [Function],
        "type": "@@query/REQUEST_ASYNC",
        "unstable_preDispatchCallback": undefined,
        "update": Object {
          "posts": [Function],
        },
        "url": "https://jsonplaceholder.typicode.com/posts",
      },
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "dispatch": [Function],
        "isLoading": true,
        "posts": null,
        "query": Object {
          "body": undefined,
          "force": undefined,
          "meta": undefined,
          "options": undefined,
          "queryKey": undefined,
          "retry": undefined,
          "transform": [Function],
          "type": "@@query/REQUEST_ASYNC",
          "unstable_preDispatchCallback": undefined,
          "update": Object {
            "posts": [Function],
          },
          "url": "https://jsonplaceholder.typicode.com/posts",
        },
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "context": Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
    },
  },
}
`;
